thrust/system/tbb/detail/scan.inl

```diff
@@ -194,32 +202,16 @@ template<typename InputIterator,
                                 OutputIterator result,
                                 BinaryFunction binary_op)
 {
-  // the pseudocode for deducing the type of the temporary used below:
-  // 
-  // if BinaryFunction is AdaptableBinaryFunction
-  //   TemporaryType = AdaptableBinaryFunction::result_type
-  // else if OutputIterator is a "pure" output iterator
-  //   TemporaryType = InputIterator::value_type
-  // else
-  //   TemporaryType = OutputIterator::value_type
-  //
-  // XXX upon c++0x, TemporaryType needs to be:
-  // result_of_adaptable_function<BinaryFunction>::type
-  
   using namespace thrust::detail;
 
   typedef typename eval_if<
-    has_result_type<BinaryFunction>::value,
-    result_type<BinaryFunction>,
-    eval_if<
-      is_output_iterator<OutputIterator>::value,
-      thrust::iterator_value<InputIterator>,
-      thrust::iterator_value<OutputIterator>
-    >
+    is_output_iterator<OutputIterator>::value,
+    thrust::iterator_value<InputIterator>,
+    thrust::iterator_value<OutputIterator>
```

Comment:
@jaredhoberock @brycelelbach @griwes I'm looking for your thoughts on this particular change. This was made to support this unit test:

```
    // float -> float with plus<int> operator (int accumulator)
    thrust::inclusive_scan(float_input.begin(), float_input.end(), float_output.begin(), thrust::plus<int>());
    ASSERT_EQUAL(float_output[0],  1.5);
    ASSERT_EQUAL(float_output[1],  3.0);
    ASSERT_EQUAL(float_output[2],  6.0);
    ASSERT_EQUAL(float_output[3], 10.0);
```
(https://github.com/thrust/thrust/blob/master/testing/scan.cu#L261-L266)

Note that this is doing an `inclusive_scan` with `float` input/output, but using `thrust::plus<int>` as the binary operator. The current version of the test expects the output to be at integer precision, except for the first element, which is `1.5`. This is how the CUDA backend behaves.

According to the comments here (deleted in this PR), this should be `1.0`, since the partial sums should have a valuetype matching the return type of the binary functor (in this case, `int`). This is how the TBB backend is currently behaving (until this PR).

Looking at the history of this test, it looks a lot like the expected value was changed arbitrarily instead of fixing a bug in the `cuda_cub` backend:
https://github.com/thrust/thrust/commit/3b6b107b940bc32e0c5701200f2347d6a03975ba#diff-95dc21c3de5f6e247592a56e50f1ea95

Even the comment there specifies that the accumulator should be `int`.

Currently this PR changes the TBB backend to match the expected test result of `1.5`, but now I'm wondering if we should instead be fixing all of the backends to return `1.0` for the first value of this test's output, since that appears to be the original intention of this unit test.

How do ya'll view this edge case? I see three possible takes:

a) Inclusive scan doesn't take an `initial_value`, so the first element should not have `binary_functor` applied to it. The correct value for the first output element is `1.5`.

b) The `binary_functor` returns `int`, so that type should be used to accumulate partial sums and the output should be at `int` precision. The correct value is `1.0`.

c) This edge case is silly and ill-defined, GIGO. We should remove this test case and say the result is implementation defined.

Reply:
As per [P0571](https://wg21.link/P0571), the intermediate type should be inferred from the input iterator, so it should be `float` here.

Reply:
Opened #1170 to track this.

Reply:
This part of the patch has been taken over by #1176.